lib/bamboo/mailer.ex

```diff
@@ -200,45 +276,49 @@ defmodule Bamboo.Mailer do
   end
 
   defp validate_and_normalize(email, adapter) do
-    email |> validate(adapter) |> normalize_addresses
+    case validate(email, adapter) do
+      :ok -> {:ok, normalize_addresses(email)}
+      error -> error
+    end
   end
 
   defp validate(email, adapter) do
-    email
-    |> validate_from_address
-    |> validate_recipients
-    |> validate_attachment_support(adapter)
+    with :ok <- validate_from_address(email),
+         :ok <- validate_recipients(email),
+         :ok <- validate_attachment_support(email, adapter) do
+      :ok
+    end
   end
 
-  defp validate_attachment_support(%{attachments: []} = email, _adapter), do: email
+  defp validate_attachment_support(%{attachments: []} = _email, _adapter), do: :ok
 
-  defp validate_attachment_support(email, adapter) do
+  defp validate_attachment_support(_email, adapter) do
     if Code.ensure_loaded?(adapter) && function_exported?(adapter, :supports_attachments?, 0) &&
          adapter.supports_attachments? do
-      email
+      :ok
     else
-      raise "the #{adapter} does not support attachments yet."
+      {:error, "the #{adapter} does not support attachments yet."}
     end
   end
 
   defp validate_from_address(%{from: nil}) do
-    raise Bamboo.EmptyFromAddressError, nil
+    {:error, %Bamboo.EmptyFromAddressError{}}
   end
 
   defp validate_from_address(%{from: {_, nil}}) do
-    raise Bamboo.EmptyFromAddressError, nil
+    {:error, %Bamboo.EmptyFromAddressError{}}
```

Comment:
I like that these changes result in more idiomatic elixir, returning the error tuple